### DISCUSSION TITLE
Best config for tacotron2 training

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#65] [Archived]
<USER>>>> geneing<br/> <DATE> [July 14, 2019, 4:31am]<br/> <br/> What is the best config for<br/> tacotron2 training. I see that master branch config_tacotron2.json is<br/> different from the one that comes with the [latest pretrained Tacotron2<br/> model](https://drive.google.com/open?id=1FJRjGDAqWIyZRX4CsppaIPEW8UWXCWzF). *slash* <br/> In particular what are better choices:<br/> <br/> 1. attention_norm: sigmoid vs softmax<br/> 2. prenet_type: original vs bn<br/> 3. loss_masking: true vs false<br/> 4. enable_eos_bos_chars: false vs true<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/best-config-for-tacotron2-training]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [July 16, 2019, 9:06am]<br/> <br/> There is no better choice, it is all about the dataset. But it is better<br/> to start with the original Tacotron paper. So it means;<br/> <br/> ![](https://cdn-production.discourse.mozilla-community.org/letter_avatar_proxy/v4/letter/g/bbce88/40.png){.avatar<br/> <br/> > - attention_norm: softmax<br/> > - prenet_type: original<br/> > - loss_masking: true<br/> > - enable_eos_bos_chars: false<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> thllwg<br/> <DATE> [June 23, 2020, 2:33pm]<br/> <br/> I would like to open a discussion about the<br/> [config.json](https://github.com/mozilla/TTS/blob/master/config.json)<br/> file included in the master branch. While questions about a '*best*'<br/> configuration may not be answered universally, a consistent exemplary<br/> configuration provides a solid foundation for individual experiments.<br/> <br/> Towards this goal, I propose to remove or change some confusing elements<br/> in the current config file. Right now, an exemplary configuration for a<br/> Tacotron2 training with LJSpeech is indicated there. This makes sense<br/> considering the 'Collaborative Experimentation Guide' advertised in the<br/> README, that equally advocates to use LJSpeech for experiments. However,<br/> there seem to be a confusing inconsistency between some of the<br/> parameters set and their respective comments.<br/> <br/> Consider for instance:<br/> <br/> `'do_trim_silence': true,// enable trimming of slience of audio as you load it. LJspeech (false), TWEB (false), Nancy (true)` *slash* <br/> If the comment proposes not to use `do_trim_silence` with LJspeech, the<br/> parameters value should be `false`.<br/> <br/> A second example:<br/> <br/> `'attention_norm': 'sigmoid', // softmax or sigmoid. Suggested to use softmax for Tacotron2 and sigmoid for Tacotron.` *slash* <br/> Set `attention_norm` to softmax, or indicate why sigmoid should be used<br/> if Tacotron2 is trained with LJSpeech.<br/> <br/> In my opinion, the problem is not only that there seems to be a mismatch<br/> between the parameter value and the comment, but also that it becomes<br/> unclear whether the values of other parameters are already adapted to<br/> LJSpeech or require further changes. While the inconsistency is obvious<br/> in the above examples, it may be the case in others as well, but less<br/> obviously.<br/> <br/> Independent of the specific training set, this comment does not reflect<br/> the source code:<br/> <br/> 'sample_rate': 22050, // DATASET-RELATED: wav sample-rate. If different than the original data, it is resampled.<br/> <br/> Yet, the audio is not resampled, see<br/> [ *slash* #405](https://github.com/mozilla/TTS/pull/405). Instead, maybe<br/> recommend to resample before training or remove that part of the<br/> comment?<br/> <br/> I understand that - as TTS is under active development - knowledge and<br/> best practices change frequently, so keeping a consistent, up-to-date<br/> config file might be difficult. Alternatively, maybe there could be a<br/> section in the wiki devoted to lessons learned regarding<br/> Model-Dataset-Configurations and exemplary configurations?<br/> <br/> I'm happy to make a PR with regards to the identified parts above.<br/> However, maybe my understanding of the whole situation is not on point,<br/> so I'm happy to hear your thoughts.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [June 23, 2020, 10:48pm]<br/> <br/> This seems worth a look - I agree that initially I ran into a few of the<br/> points you raise (eg with the softmax comment being recommended for<br/> Tacotron2 even though most of the config files which appear to be using<br/> Tacotron2 generally go against that comment!)<br/> <br/> I don't want this to distract from the<br/> raises above, but one related<br/> point has been in the back of my mind for a while, which I figured I'd<br/> mention here for general comment: **might there be some easier / better<br/> option for handling config files here?** *(can move this to another<br/> thread though if people prefer though?)*<br/> <br/> I haven't looked at many other projects on this point specifically but<br/> that seems like it's worth a go doing some research rather than<br/> reinventing the wheel.<br/> <br/> **Q.** If there was interest in this area, what features would people<br/> most value?<br/> <br/> I saw a library recently that claims to handle layered config files<br/> (plus a few other things). It's called<br/> [Confuse](https://github.com/beetbox/confuse) - I haven't used it, so<br/> can't give a personal testimony but it has a decent(ish) number of stars<br/> ( *slash* ~218 as of June 2020).<br/> <br/> What seems appealing to me about being able to layer the configs is we<br/> could have a base config that includes all the core settings, and then<br/> for a broad category (eg Tacotron vs Tacotron2) you have another layer<br/> config - but because the library is adding the configs together, you<br/> just need the key differences. And then finally with the experiments and<br/> people's individual efforts, you would have a final layer - which in<br/> most cases would be very sparse but it would also help to make it clear<br/> if there were custom/newly tried settings.<br/> <br/> Not sure how necessary this would be, but by encouraging the putting of<br/> things like the specific folder paths people use in a final personal<br/> layer (or pointing it finally to a *slash* ~/.config file, [as Confuse can<br/> do](https://github.com/beetbox/confuse#what-it-does)) it means that<br/> often there might nothing (*or almost nothing*) to edit when you get the<br/> latest updates from the repo.<br/> <br/> This would add some initial complexity but sounds like it might pay off<br/> with greater clarity for new users and some simplicity for the more<br/> experienced users too. With the comments, these could then naturallly<br/> sit in the relevant places (*eg something that was advised to be used<br/> with Tacotron2 might just be commented on in the Tacotron2 layer config<br/> file*)<br/> <br/> I'd probably suggest that for the core settings we'd then mainly just<br/> have comments that indicate allowed values or the widest limits that<br/> would ever be useful (but that's just an idea)<br/> <br/> Confuse can also handle config value validation (that hasn't typically<br/> caught me out much but I've had the odd typo I could've avoided!)<br/> <br/> Of course if others have alternative suggestions, or simply think this<br/> isn't a wise idea *(either right now or ever!)*, do please say<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [June 24, 2020, 10:07am]<br/> <br/> thx for pointing the<br/> conflicts. I just overlooked the comments as I experiment with the<br/> values and getting better results. I'll update them regarding the latest<br/> config.json. Or feel free to send a PR. It is really hard to keep things<br/> consistent as I run experiments every day. However, config.json in the<br/> TTS repo is always the best config I used so far even if there are such<br/> inconsistencies.<br/> <br/> Regarding the whole config situation, I'd favor keeping config files as<br/> simple as possible with fewer branching. That would be easier to<br/> maintain and distribute the models.<br/> <br/> Probably one approach to consider is to release recipes (as in Kaldi)<br/> per dataset with a particular config of the best model we trained so far<br/> for that dataset. That would at least ease the burden for a novice user.<br/> <br/> I'll also check Confuse.<br/> Maybe we can replicate its behaviour internally without introducing a<br/> new dependency.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [June 24, 2020, 2:29pm]<br/> <br/> I think seperating model configuration (layer configurations and model<br/> specific settings) from the rest is a good alternative since as we add<br/> new models to TTS ( as I do it now), it gets harder to accommodate all<br/> of the models in a single config.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> mardan<br/> <DATE> [June 28, 2020, 1:41pm]<br/> <br/> What about creating a separate repo which maintains an universal<br/> standard for configs related to TTS (and ASR)? So, anyone who is working<br/> on a text-to-speech project can start his configuration based on this<br/> standard.<br/> <br/> It defines unique names for common parameters required for data<br/> processing and well know models. It also suggests defaults. Maybe this<br/> makes things more consistent, like 'do_trim_silence' in TTS vs<br/> 'trim_silence' in ESPNet.<br/> <br/> This repo may also maintain a best performed config for each model with<br/> regard to datasets.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [July 7, 2020, 4:05pm]<br/> <br/> A separate repo with dataset specific runs would be a nice solution. So<br/> each dataset can have a folder and people just need to send a shell<br/> script downloading and installing everything (dataset, right TTS commit<br/> etc.) and run training.<br/> <br/> [Archived Post]<br/>
